File: 05 SWEA/D2/Python/Prob1859.py

"""
title  : 1859. 백만 장자 프로젝트 (D2)
time   : 1188ms
memory : 208092kb
"""

def solution():

    T = int(input())
    for tc in range(1, T+1):

        N = int(input())
        price = list(map(int, input().split()))

        maxPrice = price[-1]
        answer = 0

        for p in price[::-1]:

            if (p < maxPrice):
                answer += (maxPrice - p)
            else:
                maxPrice = p

        print(f'#{tc} {answer}')    

solution()

File: 05 SWEA/D2/Java/src/Prob1859.java

/**
* title  : 1859. 백만 장자 프로젝트 (D2)
* time   : 493ms
* memory : 136140kb
*/

import java.io.*;
import java.util.*;
 
public class Prob1859 {
    
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());
        for (int t = 1; t <= T; t++) {

            int day = Integer.parseInt(br.readLine());
            int[] price = new int[day];

            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; i < day; i++) {
                price[i] = Integer.parseInt(st.nextToken());
            }

            long answer = 0L;
            int maxPrice = price[day-1];
            for (int i = day-1; i >= 0; i--) {
                
                if (price[i] < maxPrice) {
                    answer += (maxPrice - price[i]);
                } else {
                    maxPrice = price[i];
                }
            }

            sb.append("#" + t + " ").append(answer).append("\n");
        }

        System.out.println(sb);
    }
}
